rpc/internal/collector.service.go

```diff
@@ -190,54 +205,107 @@ func (sc *streamsCoordinator) getStream(name string) (string, *streamGroup, erro
 	return id, group, nil
 }
 
-// addEvent writes the given event to the collector from the given stream. If
-// the stream does not exist an error is returned. Note that this function
-// blocks until all streams in the group have an entry in the heap, at which
-// point the timestamp can be guaranteed.
+// addEvent writes the given event to the stream's buffer. If the time since
+// the last flush has surpassed the flush interval, the flush method is called.
+// If the stream does not exist or is already closed an error is returned.
 func (sg *streamGroup) addEvent(ctx context.Context, id string, event *events.Performance) error {
 	sg.mu.Lock()
+	defer sg.mu.Unlock()
+
+	if sg.catcher.HasErrors() {
+		return sg.catcher.Resolve()
+	}
 
-	errChan, ok := sg.streams[id]
+	stream, ok := sg.streams[id]
 	if !ok {
-		sg.mu.Unlock()
-		return errors.Errorf("stream %s does not exist in this stream group", id)
+		return errors.Errorf("stream '%s' does not exist in this stream group", id)
+	}
+	if stream.closed {
+		return errors.Errorf("stream '%s' already closed", id)
 	}
 
-	sg.eventHeap.SafePush(&performanceHeapItem{errChan: errChan, event: event})
-	if sg.eventHeap.Len() >= len(sg.streams) {
-		item := sg.eventHeap.SafePop()
-		go sg.sendError(item)
+	if stream.inHeap {
+		stream.buffer <- event
+		return nil
 	}
 
-	sg.mu.Unlock()
+	var e *events.Performance
 	select {
-	case <-ctx.Done():
-		return errors.Wrap(ctx.Err(), "context canceled while adding event")
-	case err := <-errChan:
-		return err
+	case e = <-stream.buffer:
+		stream.buffer <- e
```

Comment:
It seems like this has the effect of taking the head of the channel and putting it at the back. Is this intentional or could this cause problems if we assume the stream buffer is not ordered by timestamp? Alternatively, would it be sufficient to just check the length of `stream.buffer` rather than dequeue/enqueue the event in the buffer?

Reply:
Yeah it was supposed to be `stream.buffer <- event`. But I think I can actually just remove this entirely, there should never be a case where there the stream is not in the heap but there is something in the buffer (thanks to the flush method which dequeues from the corresponding stream after popping an item off the heap).